src/libmeasurement_kit/ooni/captive_portal.cpp

```diff
@@ -0,0 +1,231 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "private/common/parallel.hpp"
+#include "private/common/utils.hpp"
+#include "private/ooni/constants.hpp"
+#include "private/ooni/utils.hpp"
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+//
+// Implementation note:
+// all but one of the current vendor tests are HTTP tests.
+// these check either a status code or the body content.
+//
+
+typedef std::map<std::string, std::string> input_t; // Syntactic sugar
+
+static std::vector<input_t> gen_http_inputs() {
+    std::vector<input_t> is;
+    input_t i;
+    i["name"] = "Microsoft HTTP";
+    i["url"] = "http://www.msftncsi.com/ncsi.txt";
+    i["body"] = "Microsoft NCSI";
+    is.push_back(i);
+    i.clear();
+    i["name"] = "Apple HTTP 1";
+    i["url"] = "http://captive.apple.com";
+    i["body"] = "<HTML><HEAD><TITLE>Success</TITLE></HEAD><BODY>Success</"
+                "BODY></HTML>\n";
+    i["ua"] = "CaptiveNetworkSupport/1.0 wispr";
+    is.push_back(i);
+    i.clear();
+    i["name"] = "Apple HTTP 2";
+    i["url"] = "http://captive.apple.com/hotspot_detect.html";
```

Comment:
I don't know. I defer to @joelanders. (I asked him to document the process with which he obtained all this information regarding captive portal detection in the specification of the test.)